libvast/vast/fbs/table_builder.hpp

```diff
@@ -0,0 +1,113 @@
+/******************************************************************************
+ *                    _   _____   __________                                  *
+ *                   | | / / _ | / __/_  __/     Visibility                   *
+ *                   | |/ / __ |_\ \  / /          Across                     *
+ *                   |___/_/ |_/___/ /_/       Space and Time                 *
+ *                                                                            *
+ * This file is part of VAST. It is subject to the license terms in the       *
+ * LICENSE file found in the top-level directory of this distribution and at  *
+ * http://vast.io/license. No part of VAST, including this file, may be       *
+ * copied, modified, propagated, or distributed except according to the terms *
+ * contained in the LICENSE file.                                             *
+ ******************************************************************************/
+
+#pragma once
+
+#include "vast/detail/narrow.hpp"
+#include "vast/fbs/utils.hpp"
+#include "vast/fwd.hpp"
+
+#include <flatbuffers/flatbuffers.h>
+
+#include <type_traits>
+
+namespace vast::fbs {
+
+/// Movable builder for types that wrap a FlatBuffers table.
+/// @tparam Table The wrapped FlatBuffers table to build.
+template <class Table>
+class table_builder {
+public:
+  // -- types and constants ----------------------------------------------------
+
+  using table_type = Table;
+  using derived_type = typename table_type::derived_type;
+  using root_type = typename table_type::root_type;
+  using offset_type = typename flatbuffers::Offset<root_type>;
+
+  /// A pointer to the function that returns the underlying FlatBuffers table's
+  /// file identifier.
+  inline static constexpr auto file_identifier = table_type::file_identifier;
+
+  // -- constructors, destructors, and assigmnent operators --------------------
+
+  /// Construct a builder with an initial buffer size.
+  /// @param initial_size The initial buffer size in Bytes.
```

Comment:
I'm not really convinced that there's a need for this class: As far as I can tell, its only purpose is to call `builder_.Finish()` with the correct file identifier, but in a non-generic class one would call `builder_.FinishSegment()` and this would be even safer because flatbuffers code will automatically select the correct identifier based on the information in the flatbuffer schema.

As a tradeoff, we have a virtual base class and a hard-to-describe control flow (i.e. user must call `segment_builder::add()` followed by `table_builder::finish()`, and the `create()` method directly accesses the builder from the base class)

And furthermore, it seems like this would be hard to compose, ie. how would I be able to create a non-root level segment flatbuffer as part of some existing builder?

Reply:
> I'm not really convinced that there's a need for this class: As far as I can tell, its only purpose is to call `builder_.Finish()` with the correct file identifier, but in a non-generic class one would call `builder_.FinishSegment()` and this would be even safer because flatbuffers code will automatically select the correct identifier based on the information in the flatbuffer schema.
>
> As a tradeoff, we have a virtual base class and a hard-to-describe control flow (i.e. user must call `segment_builder::add()` followed by `table_builder::finish()`, and the `create()` method directly accesses the builder from the base class)


We agreed that this is a stylistic matter mostly. Putting this off until we agreed on all other tasks.

> And furthermore, it seems like this would be hard to compose, ie. how would I be able to create a non-root level segment flatbuffer as part of some existing builder?

We agreed that the table_builder should probably be passed the FlatBuffers builder externally.